compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1362 @@
+      case SeqLiteral(elems, elemtpt) =>
+        evalExprs(elems, thisV, klass).join
+
+      case Inlined(call, bindings, expansion) =>
+        evalExprs(bindings, thisV, klass)
+        eval(expansion, thisV, klass)
+
+      case Thicket(List()) =>
+        // possible in try/catch/finally, see tests/crash/i6914.scala
+        Bottom
+
+      case vdef : ValDef =>
+        // local val definition
+        val rhs = eval(vdef.rhs, thisV, klass)
+        val sym = vdef.symbol
+        initLocal(thisV.asInstanceOf[Ref], vdef.symbol, rhs)
+        Bottom
+
+      case ddef : DefDef =>
+        // local method
+        Bottom
+
+      case tdef: TypeDef =>
+        // local type definition
+        Bottom
+
+      case _: Import | _: Export =>
+        Bottom
+
+      case tpl: Template =>
+        init(tpl, thisV.asInstanceOf[Ref], klass)
+
+      case _ =>
+        report.error("[Internal error] unexpected tree: " + expr + "\n" + Trace.show, expr)
+        Bottom
+  }
+
+  /** Handle semantics of leaf nodes
+   *
+   * For leaf nodes, their semantics is determined by their types.
+   *
+   * @param tp      The type to be evaluated.
+   * @param thisV   The value for `C.this` where `C` is represented by `klass`.
+   * @param klass   The enclosing class where the type `tp` is located.
+   * @param elideObjectAccess Whether object access should be omitted.
+   *
+   * Object access elission happens when the object access is used as a prefix
+   * in `new o.C` and `C` does not need an outer.
+   */
+  def evalType(tp: Type, thisV: Value, klass: ClassSymbol, elideObjectAccess: Boolean = false): Contextual[Value] = log("evaluating " + tp.show, printer, (_: Value).show) {
+    tp match
+      case _: ConstantType =>
+        Bottom
+
+      case tmref: TermRef if tmref.prefix == NoPrefix =>
+        val sym = tmref.symbol
+        if sym.is(Flags.Package) then
+          Bottom
+        else if sym.owner.isClass then
+          // The typer incorrectly assigns a TermRef with NoPrefix for `config`,
+          // while the actual denotation points to the symbol of the class member
+          // instead of the parameter symbol for the primary constructor.
+          //
+          // abstract class Base(implicit config: Int)
+          // case class A(x: Int)(implicit config: Int) extends Base
+          evalType(sym.termRef, thisV, klass, elideObjectAccess)
+        else
+          readLocal(thisV, sym)
+
+      case tmref: TermRef =>
+        val sym = tmref.symbol
+        if sym.isStaticObject then
+          if elideObjectAccess then
+            ObjectRef(sym.moduleClass.asClass)
+          else
+            accessObject(sym.moduleClass.asClass)
+        else
+          val value = evalType(tmref.prefix, thisV, klass)
+          select(value, tmref.symbol, tmref.prefix)
+
+      case tp @ ThisType(tref) =>
+        val sym = tref.symbol
+        if sym.is(Flags.Package) then
+          Bottom
+        else if sym.isStaticObject && sym != klass then
```

Comment:
Is it possible that `sym == klass` and `sym.isStaticObject`? If no, consider a comment. If yes, do we just not need to do anything in that case? Consider a comment to confirm this.

Reply:
Yes, that is possible --- sometimes typer use `this` type to refer to an object outside its definition.